src/services/mapCode.ts

```diff
@@ -0,0 +1,366 @@
+import {
+    Block,
+    ClassElement,
+    ClassLikeDeclaration,
+    createSourceFile,
+    factory,
+    FileTextChanges,
+    forEach,
+    forEachChild,
+    formatting,
+    hasJSDocNodes,
+    InterfaceDeclaration,
+    isBlock,
+    isClassElement,
+    isClassLike,
+    isNamedDeclaration,
+    isSourceFile,
+    isSourceFileJS,
+    isTypeElement,
+    LanguageServiceHost,
+    NamedDeclaration,
+    Node,
+    NodeArray,
+    Program,
+    ScriptTarget,
+    SourceFile,
+    SourceMapper,
+    Statement,
+    SyntaxKind,
+    sysLog,
+    textChanges,
+    TextSpan,
+    TypeElement,
+    UserPreferences,
+} from "./_namespaces/ts";
+import {
+    ChangeTracker,
+} from "./textChanges";
+
+/** @internal */
+export function mapCode(
+    mappings: { sourceFile?: SourceFile; contents: string[]; focusLocations?: TextSpan[][]; }[],
+    _program: Program,
+    host: LanguageServiceHost,
+    formatContext: formatting.FormatContext,
+    preferences: UserPreferences,
+    _sourceMapper: SourceMapper,
+    _updates?: FileTextChanges[],
+): FileTextChanges[] {
+    try {
+        return textChanges.ChangeTracker.with(
+            { host, formatContext, preferences },
+            changeTracker => {
+                mappings.forEach(({ sourceFile, contents, focusLocations }) => {
+                    if (!sourceFile) {
+                        // TODO(@zkat): uhhh... do something about file-less mappings. Not supported for now.
+                        return;
+                    }
+                    const parsed = contents.map(parse);
+                    for (const nodes of parsed) {
+                        placeNodeGroup(sourceFile, changeTracker, nodes, focusLocations);
+                    }
+                });
+            },
+        );
+    }
+    catch (e) {
+        sysLog(`mapCode: ${e}`);
+        return [];
+    }
+}
+
+/**
+ * Tries to parse something into either "top-level" statements, or into blocks
+ * of class-context definitions.
+ * @param content
+ * @returns
+ */
+function parse(content: string): NodeArray<Node> {
+    // We're going to speculatively parse different kinds of contexts to see
+    // which one makes the most sense, and grab the NodeArray from there. Do
+    // this as lazily as possible.
+    const nodeKinds = [
+        { parse: () => createSourceFile("__mapcode_content_nodes.ts", content, ScriptTarget.Latest, /*setParentNodes*/ true), body: (sf: SourceFile) => sf.statements },
+        { parse: () => createSourceFile("__mapcode_class_content_nodes.ts", `class __class {\n${content}\n}`, ScriptTarget.Latest, /*setParentNodes*/ true), body: (cw: SourceFile) => (cw.statements[0] as ClassLikeDeclaration).members },

```

Comment:
I think the line ending to use depends on the project configuration.

Reply:
that doesn't really matter in this case: as long as it's at all recognized as a line ending, it'll be fine. We throw away the surrounding class anyway.